### ISSUE TITLE
Unexpected 401 due to AuthenticationCompletionException is thrown in OIDC Authorization Code Flow

### ISSUE THREAD

**Issue by NicoNes** [quarkusio/quarkus#27348]
### Describe the bug

Hi guys,

I have an application where Quarkus OIDC Extension is enabled to protect  HTTP endpoints using OIDC Authorization Code Flow .

**_Case 1:_**

- Step 1:  I open a tab in my web browser and I try to reach a protected endpoint. I'm redirected to the OIDC provider login form to provide my credential **=> OK**
-  Step 2: I don't fill up the form and I open another tab in my web browser to reach, the same or any other protected endpoint. A 401 is returned due to AuthenticationCompletionException **=> KO** https://github.com/quarkusio/quarkus/blob/9381f2ef1b15453f4cbc6b9819478379e03962d0/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java#L131

I was expecting to be redirected to the OIDC provider login form to provide my credential.

**_Case 2:_**

- Step 1: I'm successfully  authenticated in my webapp using OIDC Authorization Code Flow.
- Step 2: I open another tab in my web browser to browse another page of my webapp. **=> OK**
- Step3: I leave for a long lunch break and both my webapp session cookie and my OIDC provider session cookie expires. **=> OK**
- Step4: After the break, in both tabs in my web browser, my webapp frontend automatically tries to reach "concurrently" a protected endpoint. **In both tabs** I'm redirected to the OIDC provider login form to provide my credential **=> OK**
-  Step 5: I fill up the form in tab 1 and a 401 is returned due to AuthenticationCompletionException **=> KO** https://github.com/quarkusio/quarkus/blob/9381f2ef1b15453f4cbc6b9819478379e03962d0/extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java#L131

I was expecting to be successfully authenticated and redirected to the protected endpoint.

**_Proposal:_**

According to me:
 
- In case 1, the error occurs because state cookie `q_auth` exists but there is no `state` query param in the current URI.
My proposal here would be first to invoke method `CodeAuthenticationMechanism.processRedirectFromOidc(...)` only when the current URI is the OIDC `redirect_uri` instead of doing it for every incoming request with a the state cookie `q_auth`.
So: 
  - when the current URI is not the OIDC `redirect_uri`, a 302 should be send with a fresh state cookie named like `q_auth_{state}`  where `{state}` is the value of the `state` query param sent to the OIDC provider. 
  - when the current URI is the OIDC `redirect_uri`, the method  `CodeAuthenticationMechanism.processRedirectFromOidc(...)` should be invoked and the `state` check should be performed by checking at least if any request cookie `q_auth_{state}` exists. 

- In case 2, the error occurs because the state cookie `q_auth` created by tab1 has been overriden by the one created by tab2 leading to a value mismatch when performing `state` check.
  The proposal I made for case 1 with the dedicated  `q_auth_{state}` cookie allows to fix this case too.

WDYT ?

Thanks

-Nicolas

**Comment by sberyozkin** [comment]
Hi @NicoNes Thanks for this report, however I think this issue is invalid and I don't agree with it being qualified as a bug.
Checking the `state` is a prerequisite which can't be avoided for security reasons - if the state cookie exists but there is no state query param in the current URI to match it then it is 401. It must happen for every request.
I don't see a reason why we should weaken this check to support a case where one gets a provider challenge on the redirect but instead of completing it opens a new tab and tries to access Quarkus again...
If necessary you should open the tab in the Incognito mode - that should address your issue too
Thanks

**Comment by NicoNes** [comment]
Hi @sberyozkin ,

> Checking the state is a prerequisite which can't be avoided for security reasons - if the state cookie exists but there is no state query param in the current URI to match it then it is 401. It must happen for every request.

The `state` check is important for security reason to prevent CSRF I know and we both agree with that. 
But this check is only relevant when OIDC provider send Authorization code back on the `redirect_uri`.
I suppose you already know the OAuth2 RFC but it is written in it :

https://www.rfc-editor.org/rfc/rfc6749#section-4.1.1.
> state
>          RECOMMENDED.  An opaque value used by the client to maintain
>          state between the request and **callback**.

The expected usage of state is even described in the RFC:

https://www.rfc-editor.org/rfc/rfc6749#section-10.12

>    The client MUST implement CSRF protection for **its redirection URI**.
>    This is typically accomplished by requiring any request sent to the
>    redirection URI endpoint to include a value that binds the request to
>    the user-agent's authenticated state...

Most of the webapp I've been using behave like that that's why Quarkus behavior on this point surprised me a bit. Try it on gitlab or spotify for example you will see my point.

About case 2, it's a normal situation to open multiple tabs when using a webapp and being disconnected on each of them after a long break.  How can my user know which tab to use to fill up the OIDC provider login form to avoid state mismatch ?

Thanks

**Comment by sberyozkin** [comment]
@NicoNes I've had a look for a few hours today into this issue. Lets focus on the case 1 for now.

Can you explain to me again, why as a user, after being redirected to the OIDC authorization endpoint and being challenged with the authentication dialog, I'd open a new tab and try to access the same Quarkus application and start a new authentication process ? What is the idea/use case behind it ?

Note in Quarkus we do not have a notion of a single Quarkus application maintaining multiple sessions for the same user for different parts of this application, with each part requiring a separate authentication for the same user. So once a user has authenticated when accessing a protected `/a` Quarkus path, the user will be able to access the protected `/b` path.

So right now it does not make sense to me, trying to support such a multi-tab experience - start in one tab, then another one while not finishing the 1st tab flow, etc.

What am I missing ?

**Comment by NicoNes** [comment]
Hi @sberyozkin ,

It's actually case 2 that leads me to case 1, but lets try to focus on case 1 first.

> Can you explain to me again, why as a user, after being redirected to the OIDC authorization endpoint and being challenged with the authentication dialog, I'd open a new tab and try to access the same Quarkus application and start a new authentication process, without completing the authentication challenge in the current tab ? What is the idea/use case behind it ?

Well, as I said it's case 2 that leads me to case 1 so I have no specific use case behind case 1. 

However,  here is a real life flow that can lead to case 1 :

- Open tab 1 on YOUTUBE to listen music while working.
- Open tab 2 on quarkus app A to start working but without filling up the authentication form because I was interrupted by my boss calling me to come at his desk.
- Come back to my desk and click on a link that my boss sent me on SKYPE that opens tab 3 on quarkus app A. => Error

> So right now it does not make sense to me, trying to support such a multi-tab experience - start in one tab, then another one while not finishing the 1st tab flow, etc.

Well, to me, it's the state check on every path, when state cookie exists, that does not make sense. What is the purpose of doing that ? What is the security requirement behind this ? Why should I prevent user from opening multiple tabs on the same application without flling up the authentication form (see my use case above) ?
I as said in my other comment AFAIK this is not an OAuth2 neither an OIDC security requirement to check state on every path. The security requirement is to check state only on callback path. Many well known applications support case 1 (skype, confluence, jira, gitlab ...).

So to fix case 1 if you do not want to create mulitple state cookie as I said in my previous comments (fix for case 2), I suggest to only check state when OAuth2 callback is called. When other paths are called just renew the state cookie.

WDYT ?

 -Nicolas

**Comment by sberyozkin** [comment]
Hi @NicoNes the real world use case you've described does not warrant a serious consideration IMHO (step 2 in particular - so I'm about to enter my name and password but then decide to step outside due to some emergency - well if it is the case - lets just restart a browser and start again when I'm back ?).

What is important from my point's of view is that `quarkus-oidc` supports users in production environments. The early state cookie check allows to optimize things nicely. I've started prototyping a fix along the lines of what you have proposed and started questioning it by asking myself - why I'm doing it. Now we have an edge case to deal with, someone sends an initial request to the callback path with a custom `state` query parameter.

>Why should I prevent user from opening multiple tabs on the same application without flling up the authentication form (see my use case above) ?

IMHO this is just not a real world/production use case - so for now I'm not sure it is worth changing the code just for the sake of allowing opening multiple tabs without completing the authentication challenge.

**Comment by sberyozkin** [comment]
As far as the security consideration is concerned: where is a guarantee that someone is not provoking a redirect loop by affecting the redirect URI of the authenticated user - for example, by replacing `/callback` with `/a` ?

However the case 2 looks more realistic - I've authenticated via one of the tabs and then use a few tabs to access the Quarkus web-app. Can you clarify step 4 please,
```
Step4: After the break, in both tabs in my web browser, my webapp frontend automatically tries to reach "concurrently" a protected endpoint.
```
How can this happen ? Are you thinking about SPA/JavaScript doing some regular polling of the Quarkus web-app ?

**Comment by NicoNes** [comment]
Hi @sberyozkin

I just answer on case 1 for now to not introduce confusion and I will answer for case 2 when we are done on this one.

My answers:

>  (step 2 in particular - so I'm about to enter my name and password but then decide to step outside due to some emergency - well if it is the case - lets just restart a browser and start again when I'm back ?). 

 So you mean that when the user comes back to his desk after being interrupted for a meeting or any emergency reason, he can't click on a link somebody sent to him ? 
You mean that before clicking on the link, the user must either :
- restart is browser and lose everything he was doing previously
- or first check if the link is a link to go on "quarkus app A". If true, the user must remind that he already started the authentication process for this app, and finish it before clicking on the link right ? 

Well, any of your solution seems realistic to me. We are putting constraints on how user is supposed to use his browser or application for no reason.

> What is important from my point of view is that quarkus-oidc supports users in production environments. The early state cookie check allows to optimize things nicely. 

What kind of optimization ? To me, this early state cookie check seems Quarkus specific and has no security value for a production environment.

> I've started prototyping a fix along the lines of what you have proposed and started questioning it by asking myself - why I'm doing it. 

I'd say your doing it because it's the way the spec says to do it. And because checking state for every path has no security value.

> Now we have an edge case to deal with, someone sends an initial request to the callback path with a custom state query parameter.

Not sure to understand your point here. This is not and edge case, this is the basic case described in the spec: 
Whenever the callback path is called a check state must be perform. If the custom state query parameter does not match the one in the cookie, or if there is no cookie, the request processing must fail.

> As far as the security consideration is concerned: where is a guarantee that someone is not provoking a redirect loop by affecting the redirect URI of the authenticated user - for example, by replacing /callback with /a ?

I don't understand what you mean here, can you please explain ?

**Comment by NicoNes** [comment]
> What I care about is that quarkus-oidc is production ready and the use case you have described is not what I'd expect people do in the production, to justify the proposed code modifications. Note I'm open to having quarkus-oidc as flexible as possible but there has to be a proper justification.

Well, I think the way people use an application or their browser is very subjective and personal (you said you do not expect people to do this and I said that my users can do it. I can't say you're wrong and you can't say I'm wrong neither), that's why I try to stick to concrete things like spec, security value and also the way other website/framework works even if I sometimes others can do wrong. So if you do not care about that concrete things I'm afraid we will never be agree on this case 1.

> The edge case I had in mind is that a non-authenticated user does a first request to the callback url with a custom state query parameter.

As I said in previous comment, this is not an edge case at all. If this case happens we just have to perform the check state normally. I mean:
- if there is no state cookie, the request must fail.
- else if the state query param does not match the value in the state cookie, the request must fail.

> The user has authenticated and is being redirected back via an interposed hostile actor website - which is replacing a location header changing http://localhost:8080/callback to http://localhost:8080/a. With your proposal we ignore the fact the state cookie exists and redirect the user back - OIDC provider sees the user has already authenticated and redirects the user back etc

Well in this case : 

-  with my proposal there will be an automatic infinite redirection loop between the application and the OIDC provider but no security issue, no credentials stolen nor sent to the attacker.  The automatic loop can be mitigate/stopped with fault tolerance solution.
-  with your proposal  there will be an "manual" infinite redirection loop between the application and the OIDC provider (each time user get the 401 he will try to refresh the page etc) but no security issue, no credentials stolen nor sent to the attacker. 

So in both case we are stuck in a loop but with no security leaks and we have to deal with the fact that an hostile actor try to bother us.

Well I don't know what to say more to convince you to change the code. From my point of view my proposal is standard, spec compliant and offers more flexibility to deal with the subjective part of how we think user can use the app. But it does not introduce any/more security issue than the current one.

Let me know if you can change your mind on that else I will have to deal with that on my side trying to delete state request cookie in some kind of filters when I'm not on callback path.

Thanks

**Comment by sberyozkin** [comment]
@NicoNes It's not about whose users are wrong and whose are not. `quarkus-oidc` has been around for a while and we've seen many confirmations it works for users in production with all sort of providers. Your request is the first time of this kind (likely in fact the 2nd as I recall a similar query on Zulip).

When a user sees an authentication challenge, I do not understand, why the code has to be tweaked, to let users move away from this challenge and forgetting about it and then authenticating to the same application in another tab, as opposed to expecting the users complete this challenge. It is not really about some freedom of choice etc - but about a basic completion of the flow.

However, I'd like to keep this issue open to gather more feedback - thanks for your input so far.

@pedroigor Hey Pedro - what is your opinion.

**Comment by NicoNes** [comment]
> It's not about whose users are wrong and whose are not. quarkus-oidc has been around for a while and we've seen many confirmations it works for users in production with all sort of providers. Your request is the first time of this kind (likely in fact the 2nd as I recall a similar query on Zulip).

I'm not saying it does not work for user in production. I'm telling you it works fine but it could be more flexible to handle a case like case 1 which is a realistic and reproducible production case.  More, being that flexible will not even have security impact for `quarkus-oidc`.

> When a user sees an authentication challenge, I do not understand, why the code has to be tweaked, to let users move away from this challenge and forget about it and then authenticate to the same application in another tab, as opposed to expecting the users complete this challenge. It is not really about some freedom of choice etc - but about a basic completion of the flow.

Once again, user does not move away from challenge and open the application on another tab **on purpose**. But there is many cases where this can happen unintentionally and it would have been nice from `quarkus-oidc` to handle it.

Here is another basic application flow (the last one) closer to us. Let's assume that Github authentication is provided by `quarkus-oidc`:

- I open my browser and open tab 1 on my mail box to check my mails.
- I see that I received a notification mail from Github to let me know that you answered me.
- I click on the Github link in the notification mail that opens a tab 2 on github  challenging me for authentication.
- The phone ring preventing me from completing the challenge. (That was not my intention to not complete the challenge but I have to take the call)
- I talk with my friend on the phone which tell me to go check the email he just forwarded me.
- I go back  in tab 1 to read his email.
- I finish talking with my friend and I hang up the phone
- I'm still in tab 1 in my mail box and I notice that I had a second Github notification mail to tell me that you sent me a second anwser.
- As a normal user I click on the link in the second notification mail to read your answer and I'm not expecting to end up on an error page in tab 3 just because I should have remember that I already started an authentication flow that I did not complete in tab 2. 
The error page looks like a "punishment" for not having completing the challenge in tab 2 before taking the phone call or before clicking on the second link.

To me such behavior is frustrating for user experience and, for what it worth as an justification, I suppose that also why well known application handle this little case. Handle it has no impact on security and improve user experience so why not to do it ?

> However, I'd like to keep this issue open to gather more feedback - thanks for your input so far.

You're welcome, thanks too :+1: . Let's wait for more input.

- Nicolas

**Comment by nicolaos** [comment]
I could not have explained it better than @NicoNes .

We are currently experiencing the same issue with our periodic upgrades of dependencies.
We are now forced to stop updating the quarkus dependency on theese projects, leaving us vulnerable in case some new vulnerability arises.

The evolution of quarkus-oidc and the check on the state parameter on the first engagement of the webapp caught us off guard.
We expected a redirect to a new login process, not a 401 page with users stuck without clues.

**Comment by sberyozkin** [comment]
@nicolaos I think we should not mix the feature request like this one with the vulnerability management.

>We expected a redirect to a new login process, not a 401 page with users stuck without clues.

Fair enough, but I haven't seen so far any real use case which justifies the user forgetting about the current challenge, then opening another tab, then may be forgetting about the 2nd tab's challenge, then opening another tab, etc.

But looks like a number of users interested in this feature is growing, I see 4 requests so far incl the latest one in Zulip, so I'll try to prioritize, I've been thinking of introducing a strategy for this lax state cookie verification

**Comment by yoadey** [comment]
Thanks for looking into this and just for information: we currently have an internal testing app and this issue lead to some service calls of confused users. Just reloading the page worked normally, but that's not something the user expects to see.

**Comment by crumohr** [comment]
I'm not completely sure if it's related but run into similars problems with the `security-openid-connect-web-authentication-quickstart` for `2.15.3.Final` and `2.16.1.Final`.

See https://github.com/quarkusio/quarkus-quickstarts/issues/1237

**Comment by sberyozkin** [comment]
Trying to focus on this issue right now.

**Comment by sberyozkin** [comment]
FYI, #31079.

Please test, unfortunately I can't request a `backport to 2.16`, but it should be available in a few weeks in one of the next 3.0 alpha releases. Thanks everyone for their patience

**Comment by NicoNes** [comment]
Hi @sberyozkin,

Thanks for the improvment. I'll try to test it soon :+1:

**Comment by sberyozkin** [comment]
@nicolaos Hi, thanks for all the input, give it a try when you get a chance

**Comment by allertonm** [comment]
I came to this issue because I found that the state check with the q_auth cookie also affects any situation where you have a mixture of authenticated and unauthenticated resources. 

If you hit an unauthenticated resource, it will load correctly the first time, because the request has no `q_auth` cookie. But the `q_auth` cookie will be set on the response. So this means that if you reload the page, it will 401 because the state check fails, and even more painfully, if the page contains a link to an authenticated resource, and the user clicks on it, it will also 401 rather than perform the redirect as expected.

It's absolutely painful that we're expected to upgrade to 3.0 to get a fix for this.
